src/operator/controllers/intents_reconcilers/database/database.go

```diff
@@ -39,51 +42,99 @@ func NewDatabaseReconciler(
-			r.RecordWarningEventf(intents, ReasonApplyingDatabaseIntentsFailed, "Failed applying database intents: %s", err.Error())
+	databaseToIntents, err := r.MapDBInstanceToIntents(dbIntents)
+	if err != nil {
+		r.RecordWarningEventf(clientIntents, ReasonApplyingDatabaseIntentsFailed, "Failed applying database clientIntents: %s", err.Error())
+		return ctrl.Result{}, errors.Wrap(err)
+	}
+
+	for databaseInstance, intents := range databaseToIntents {
+		pgServerConfigs := otterizev1alpha3.PostgreSQLServerConfigList{}
+		err := r.client.List(ctx, &pgServerConfigs)
+		if err != nil {
+			r.RecordWarningEventf(clientIntents, ReasonErrorFetchingPostgresServerConfig,
+				"Error trying to fetch '%s' PostgresServerConf for client '%s'. Error: %s",
+				databaseInstance, clientIntents.GetServiceName(), err.Error())
+			return ctrl.Result{}, nil
+		}
+		pgServerConf, err := findMatchingPGServerConfForDBInstance(databaseInstance, pgServerConfigs)
+		if err != nil {
+			r.RecordWarningEventf(clientIntents, ReasonMissingPostgresServerConfig,
+				"Could not find matching PostgreSQLServerConfig. Error: %s", err.Error())
+			return ctrl.Result{}, nil
+		}
+
+		pgConfigurator := databaseconfigurator.NewPostgresConfigurator(pgServerConf.Spec, r.client)
+		err = pgConfigurator.ConfigureDBFromIntents(ctx, clientIntents.GetServiceName(), clientIntents.Namespace, intents, action)
+		if err != nil {
+			r.RecordWarningEventf(clientIntents, ReasonApplyingDatabaseIntentsFailed,
+				"Failed applying database clientIntents: %s", err.Error())
 			return ctrl.Result{}, errors.Wrap(err)
 		}
-		r.RecordWarningEventf(intents, ReasonApplyingDatabaseIntentsFailed, "Failed applying database intents: %s", errMsg)
-		return ctrl.Result{}, errors.Wrap(err)
 	}
 
-	r.RecordNormalEventf(intents, ReasonAppliedDatabaseIntents, "Database intents reconcile complete, reconciled %d intent calls", len(intentInputList))
+	r.RecordNormalEventf(clientIntents, ReasonAppliedDatabaseIntents, "Database clientIntents reconcile complete, reconciled %d intent calls", len(dbIntents))
 
 	return ctrl.Result{}, nil
 }
+
+func (r *DatabaseReconciler) MapDBInstanceToIntents(intents []otterizev1alpha3.Intent) (map[string][]otterizev1alpha3.Intent, error) {
+	dbInstanceToIntents := map[string][]otterizev1alpha3.Intent{}
+	for _, intent := range intents {
+		// Name represents a database instance which is represented by a matching server config CRD
+		instanceIntents, ok := dbInstanceToIntents[intent.Name]
+		if !ok {
+			dbInstanceToIntents[intent.Name] = []otterizev1alpha3.Intent{intent}
+			continue
+		}
+		dbInstanceToIntents[intent.Name] = append(instanceIntents, intent)
+	}
+
+	return dbInstanceToIntents, nil
+}
+
+func findMatchingPGServerConfForDBInstance(
+	databaseInstanceName string,
+	pgServerConfigList otterizev1alpha3.PostgreSQLServerConfigList) (*otterizev1alpha3.PostgreSQLServerConfig, error) {
+
+	for _, conf := range pgServerConfigList.Items {
```

Comment:
can use `lo.Find`

Reply:
I know I'm considered an lo-enthusiastic in the team, it's just that it removes so much unnecessary code 😍